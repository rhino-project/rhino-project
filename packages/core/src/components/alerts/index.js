import React from 'react';
import PropTypes from 'prop-types';

import { Alert } from 'reactstrap';

export const DetailAlert = ({
  title,
  description,
  color = 'primary',
  children
}) => {
  return (
    <Alert color={color}>
      <h6>{title}</h6>
      {description && <p>{description}</p>}
      {children}
    </Alert>
  );
};

DetailAlert.propTypes = {
  children: PropTypes.node,
  color: PropTypes.string,
  description: PropTypes.node,
  title: PropTypes.string.isRequired
};

export const SuccessAlert = (props) => (
  <DetailAlert {...props} color="success" />
);

export const DangerAlert = (props) => <DetailAlert {...props} color="danger" />;
