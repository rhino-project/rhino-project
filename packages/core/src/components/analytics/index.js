export * from './IdentityAnalytics';
export * from './PageAnalytics';
