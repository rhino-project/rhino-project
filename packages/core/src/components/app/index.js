export * from './AccountMenu';
export * from './BaseOwnerSwitcher';
export * from './NotificationMenu';
export * from './Toaster';
