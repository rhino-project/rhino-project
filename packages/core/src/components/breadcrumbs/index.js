import PropTypes from 'prop-types';
import { BreadcrumbItem } from 'reactstrap';
import { useBaseOwnerPath } from '../../hooks/history';

export const BreadcrumbItemWrapper = ({ tag, to, children, ...props }) => {
  const baseOwnerPath = useBaseOwnerPath();
  const path = baseOwnerPath.build(to);

  return (
    <BreadcrumbItem tag={tag} to={path} {...props}>
      {children}
    </BreadcrumbItem>
  );
};

BreadcrumbItemWrapper.propTypes = {
  tag: PropTypes.elementType,
  to: PropTypes.string,
  children: PropTypes.node
};
