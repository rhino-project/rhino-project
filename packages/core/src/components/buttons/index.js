import PropTypes from 'prop-types';
import { useState } from 'react';
import { NavLink } from 'react-router-dom';

import { Button as RSButton } from 'reactstrap';

import { Icon } from '../icons';

export const Button = ({ loading = false, children, ...props }) => {
  if (!loading) return <RSButton {...props}>{children}</RSButton>;

  return (
    <RSButton {...props} disabled>
      <span
        className="spinner-border spinner-border-sm"
        role="status"
        aria-hidden="true"
      ></span>
      <span className="invisible">{children}</span>
    </RSButton>
  );
};

Button.propTypes = {
  children: PropTypes.node,
  loading: PropTypes.bool.isRequired
};

Button.defaultProps = {
  loading: false
};

export const IconButton = ({ children, icon, ...props }) => (
  <Button {...props}>
    <Icon
      className="me-1"
      style={{ height: '1rem', width: '1rem' }}
      icon={icon}
    />
    {children}
  </Button>
);

IconButton.propTypes = {
  children: PropTypes.node,
  icon: PropTypes.string.isRequired
};

export const OmniIconButton = ({ icon, endpoint, ...props }) => {
  const [loading, setLoading] = useState(false);

  return (
    <form method="post" action={endpoint} onSubmit={() => setLoading(true)}>
      <div className="omni-icon-btn">
        <Button
          className="omni-icon-btn__icon"
          {...props}
          type="submit"
          loading={loading}
        >
          <Icon style={{ height: '1rem', width: '1rem' }} icon={icon} />
        </Button>
      </div>
    </form>
  );
};

OmniIconButton.propTypes = {
  icon: PropTypes.string.isRequired,
  endpoint: PropTypes.string.isRequired
};

export const SubmitButton = (props) => (
  <Button {...props} color="primary" type="submit" />
);

const requiredToOrHref = (props, propName, componentName) => {
  if (!props.to && !props.href) {
    return new Error(
      `One of props 'to' or 'href' was not specified in '${componentName}'.`
    );
  }
};

export const LinkButton = (props) => {
  const tag = props.to ? NavLink : 'a';

  return <Button {...props} tag={tag} />;
};

LinkButton.propTypes = {
  to: PropTypes.string,
  href: PropTypes.string,
  toOrHref: requiredToOrHref
};

export const LinkIconButton = (props) => {
  const tag = props.to ? NavLink : 'a';

  return <IconButton {...props} tag={tag} />;
};

LinkIconButton.propTypes = {
  to: PropTypes.string,
  href: PropTypes.string,
  toOrHref: requiredToOrHref
};

export const CloseButton = (props) => <Button close {...props} />;
