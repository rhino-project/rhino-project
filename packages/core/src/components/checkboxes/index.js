import { forwardRef, useRef, useEffect } from 'react';
import { Input } from 'reactstrap';

const IndeterminateCheckboxComponent = ({ indeterminate, ...rest }, ref) => {
  const defaultRef = useRef();
  const resolvedRef = ref || defaultRef;

  useEffect(() => {
    if (resolvedRef) resolvedRef.current.indeterminate = indeterminate;
  }, [resolvedRef, indeterminate]);

  return <Input type="checkbox" innerRef={resolvedRef} {...rest} />;
};

export const IndeterminateCheckbox = forwardRef(IndeterminateCheckboxComponent);
IndeterminateCheckbox.displayName = 'IndeterminateCheckbox';
