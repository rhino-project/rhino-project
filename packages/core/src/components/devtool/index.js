import * as devtool from './RhinoDevTool';

export const RhinoDevTool =
  // eslint-disable-next-line no-undef
  process.env.NODE_ENV !== 'development'
    ? function () {
        return null;
      }
    : devtool.RhinoDevTool;
