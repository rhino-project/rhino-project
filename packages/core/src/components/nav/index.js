import classnames from 'classnames';
import PropTypes from 'prop-types';
import { NavLink as RRNavLink } from 'react-router-dom';
import { NavLink, NavItem as RSNavItem } from 'reactstrap';

import { NavIcon } from '../icons';
import { useGlobalComponent, useRoles } from '../../hooks';
import { getBaseOwnedModels, getModel, getModelIndexPath } from '../../utils';
import { useMemo } from 'react';
import { map, uniqBy } from 'lodash-es';

export const NavSection = ({
  title,
  icon,
  onIconClick,
  children,
  className
}) => {
  return (
    <div className={className}>
      {title && (
        <div className="nav-sidebar-section d-flex justify-content-between align-items-center px-0 mt-3 mb-1 text-capitalize text-gray-400">
          <span className="fw-medium">{title}</span>
          {icon && <NavIcon role="button" icon={icon} onClick={onIconClick} />}
        </div>
      )}
      <ul className="nav flex-column">{children}</ul>
    </div>
  );
};

NavSection.propTypes = {
  icon: PropTypes.string,
  title: PropTypes.node,
  children: PropTypes.node.isRequired,
  onIconClick: PropTypes.func,
  className: PropTypes.string
};

export const NavItem = ({ title, icon, extraClass, ...props }) => {
  return (
    <RSNavItem>
      <NavLink
        tag={RRNavLink}
        className={classnames(
          'd-flex',
          'flex-row',
          'align-items-center',
          'justify-content-start',
          extraClass
        )}
        {...props}
      >
        <div className="d-flex align-items-center">
          {icon && <NavIcon icon={icon} />}
          <div className={classnames({ 'ms-2': icon })}>{title}</div>
        </div>
      </NavLink>
    </RSNavItem>
  );
};

NavItem.propTypes = {
  icon: PropTypes.string,
  title: PropTypes.node.isRequired,
  to: PropTypes.string.isRequired,
  extraClass: PropTypes.string
};

const modelsRoute = (model) => getModelIndexPath(model);

export const ModelNavSectionBase = ({
  title = 'Resources',
  className,
  itemClass,
  models = null
}) => {
  const roles = useRoles();

  const fullModels = useMemo(() => {
    if (!models) {
      return getBaseOwnedModels().filter(
        (m) => m.model !== 'users_role' && m.model !== 'users_role_invite'
      );
    }

    let generatedModels = models;
    if (typeof models === 'function') {
      generatedModels = models(roles);
    }

    if (Array.isArray(generatedModels)) {
      return generatedModels.map((m) => getModel(m));
    } else if (typeof generatedModels === 'object') {
      return uniqBy(
        roles.reduce((previousValue, roleName) => {
          if (models[roleName]) {
            return [...previousValue, ...models[roleName]];
          }
          return previousValue;
        }, []),
        (model) => model
      ).map((m) => getModel(m));
    }

    console.warn('Invalid models passed to PrimaryNavigation', models);
  }, [models, roles]);

  return (
    <NavSection title={title} className={className}>
      {map(fullModels, (m) => (
        <NavItem
          key={m.model}
          title={m.pluralReadableName}
          to={modelsRoute(m)}
          icon="list"
          className="px-3"
          extraClass={itemClass}
        />
      ))}
    </NavSection>
  );
};

ModelNavSectionBase.propTypes = {
  title: PropTypes.string,
  className: PropTypes.string,
  itemClass: PropTypes.string
};

export const ModelNavSection = (props) =>
  useGlobalComponent('ModelNavSection', ModelNavSectionBase, props);
