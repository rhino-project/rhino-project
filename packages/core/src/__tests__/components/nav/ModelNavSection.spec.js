import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { render } from '@testing-library/react';
import { createWrapper } from '../../shared/helpers';
import { sharedGlobalTests } from '../../shared/sharedGlobalTests';
import { ModelNavSection } from '../../../components/nav';

vi.mock('../../../hooks', async (importOriginal) => {
  const actual = await importOriginal();
  return {
    ...actual,
    useRoles: vi.fn().mockReturnValue(['admin'])
  };
});

describe('ModelNavSection', () => {
  const Wrapper = ({ children, ...props }) => {
    const queryClient = new QueryClient();

    return (
      <MemoryRouter {...props}>
        <QueryClientProvider client={queryClient}>
          <Routes>
            <Route path="/:baseOwnerId/*" element={<>{children}</>} />
          </Routes>
        </QueryClientProvider>
      </MemoryRouter>
    );
  };

  sharedGlobalTests(ModelNavSection);

  it('renders without crashing', () => {
    const { asFragment } = render(<ModelNavSection />, {
      wrapper: createWrapper(Wrapper, {
        initialEntries: ['/1']
      })
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays models passed as string of arrays', () => {
    const { asFragment } = render(<ModelNavSection models={['user']} />, {
      wrapper: createWrapper(Wrapper, {
        initialEntries: ['/1']
      })
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays models returned from function as string of arrays', () => {
    const models = () => ['user'];
    const { asFragment } = render(<ModelNavSection models={models} />, {
      wrapper: createWrapper(Wrapper, {
        initialEntries: ['/1']
      })
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays models passed as an object', () => {
    const { asFragment } = render(
      <ModelNavSection models={{ admin: ['blog'] }} />,
      {
        wrapper: createWrapper(Wrapper, {
          initialEntries: ['/1']
        })
      }
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays models returned from function as an object', () => {
    const models = () => ['blog'];
    const { asFragment } = render(<ModelNavSection models={models} />, {
      wrapper: createWrapper(Wrapper, {
        initialEntries: ['/1']
      })
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
