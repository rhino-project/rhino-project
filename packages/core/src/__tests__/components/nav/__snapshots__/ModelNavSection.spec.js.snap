// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ModelNavSection > displays models passed as an object 1`] = `
<DocumentFragment>
  <div>
    <div
      class="nav-sidebar-section d-flex justify-content-between align-items-center px-0 mt-3 mb-1 text-capitalize text-gray-400"
    >
      <span
        class="fw-medium"
      >
        Resources
      </span>
    </div>
    <ul
      class="nav flex-column"
    >
      <li
        class="nav-item"
      >
        <a
          class="px-3 nav-link"
          href="/1/blogs"
        >
          <div
            class="d-flex align-items-center"
          >
            <svg
              class="nav-icon me-1"
              fill="currentColor"
            >
              <use
                xlink:href="/node_modules/bootstrap-icons/bootstrap-icons.svg#list"
              />
            </svg>
            <div
              class="ms-2"
            >
              Blogs
            </div>
          </div>
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`ModelNavSection > displays models passed as string of arrays 1`] = `
<DocumentFragment>
  <div>
    <div
      class="nav-sidebar-section d-flex justify-content-between align-items-center px-0 mt-3 mb-1 text-capitalize text-gray-400"
    >
      <span
        class="fw-medium"
      >
        Resources
      </span>
    </div>
    <ul
      class="nav flex-column"
    >
      <li
        class="nav-item"
      >
        <a
          class="px-3 nav-link"
          href="/1/users"
        >
          <div
            class="d-flex align-items-center"
          >
            <svg
              class="nav-icon me-1"
              fill="currentColor"
            >
              <use
                xlink:href="/node_modules/bootstrap-icons/bootstrap-icons.svg#list"
              />
            </svg>
            <div
              class="ms-2"
            >
              Users
            </div>
          </div>
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`ModelNavSection > displays models returned from function as an object 1`] = `
<DocumentFragment>
  <div>
    <div
      class="nav-sidebar-section d-flex justify-content-between align-items-center px-0 mt-3 mb-1 text-capitalize text-gray-400"
    >
      <span
        class="fw-medium"
      >
        Resources
      </span>
    </div>
    <ul
      class="nav flex-column"
    >
      <li
        class="nav-item"
      >
        <a
          class="px-3 nav-link"
          href="/1/blogs"
        >
          <div
            class="d-flex align-items-center"
          >
            <svg
              class="nav-icon me-1"
              fill="currentColor"
            >
              <use
                xlink:href="/node_modules/bootstrap-icons/bootstrap-icons.svg#list"
              />
            </svg>
            <div
              class="ms-2"
            >
              Blogs
            </div>
          </div>
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`ModelNavSection > displays models returned from function as string of arrays 1`] = `
<DocumentFragment>
  <div>
    <div
      class="nav-sidebar-section d-flex justify-content-between align-items-center px-0 mt-3 mb-1 text-capitalize text-gray-400"
    >
      <span
        class="fw-medium"
      >
        Resources
      </span>
    </div>
    <ul
      class="nav flex-column"
    >
      <li
        class="nav-item"
      >
        <a
          class="px-3 nav-link"
          href="/1/users"
        >
          <div
            class="d-flex align-items-center"
          >
            <svg
              class="nav-icon me-1"
              fill="currentColor"
            >
              <use
                xlink:href="/node_modules/bootstrap-icons/bootstrap-icons.svg#list"
              />
            </svg>
            <div
              class="ms-2"
            >
              Users
            </div>
          </div>
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`ModelNavSection > renders without crashing 1`] = `
<DocumentFragment>
  <div>
    <div
      class="nav-sidebar-section d-flex justify-content-between align-items-center px-0 mt-3 mb-1 text-capitalize text-gray-400"
    >
      <span
        class="fw-medium"
      >
        Resources
      </span>
    </div>
    <ul
      class="nav flex-column"
    >
      <li
        class="nav-item"
      >
        <a
          class="px-3 nav-link"
          href="/1/blogs"
        >
          <div
            class="d-flex align-items-center"
          >
            <svg
              class="nav-icon me-1"
              fill="currentColor"
            >
              <use
                xlink:href="/node_modules/bootstrap-icons/bootstrap-icons.svg#list"
              />
            </svg>
            <div
              class="ms-2"
            >
              Blogs
            </div>
          </div>
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`ModelNavSection > should render with global override shorthand 1`] = `
<DocumentFragment>
  <div>
    Bar
  </div>
</DocumentFragment>
`;
